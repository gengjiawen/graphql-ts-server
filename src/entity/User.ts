import { BaseEntity, BeforeInsert, Column, Entity, PrimaryGeneratedColumn } from 'typeorm'
import * as bcrypt from 'bcryptjs'
import { ObjectType, Field, ID } from 'type-graphql'

@ObjectType()
@Entity('users')
export class User extends BaseEntity {
  @Field(() => ID)
  @PrimaryGeneratedColumn('uuid')
  id: string

  @Field()
  @Column('varchar', { length: 255 })
  email: string

  @Field()
  @Column('text')
  password: string

  @Field()
  @Column('boolean', { default: false })
  confirmed: boolean

  @BeforeInsert()
  async hashPassword() {
    this.password = await bcrypt.hash(this.password, 10)
  }
}
