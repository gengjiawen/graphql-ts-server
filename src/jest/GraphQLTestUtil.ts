import axios, { AxiosResponse } from 'axios'

export function graphQLRequest(payload: string) {
  const url = process.env.TEST_HOST as string
  return axios.post(url, { query: payload })
}

export function mapErrorsMessage(response: AxiosResponse) {
  const errors: any[] = response.data.errors
  return errors.map(i => i.message)
}
