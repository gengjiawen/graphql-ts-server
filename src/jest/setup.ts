import { startServer } from '../StartServer'
import { AddressInfo } from 'net'

export default async function() {
  if (!process.env.TEST_HOST) {
    const app = await startServer()
    const { port } = app.address() as AddressInfo
    const testHost = `http://localhost:${port}`
    console.log(testHost)
    process.env.TEST_HOST = testHost
  }
}
