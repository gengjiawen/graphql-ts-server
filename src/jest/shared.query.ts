export const registerMutation = (e: string, p: string) => `
mutation {
  register(email: "${e}", password: "${p}") {
    token
  }
}`

export const loginMutation = (e: string, p: string) => `
mutation {
  login(email: "${e}", password: "${p}") {
    token
  }
}`
