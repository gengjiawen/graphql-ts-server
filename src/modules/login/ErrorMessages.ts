export const invalidLogin = 'invalid login'
export const confirmEmailError = 'please confirm your email first'
export const accountLockedError = 'account is locked'
