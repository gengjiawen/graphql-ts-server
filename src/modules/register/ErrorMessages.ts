export const duplicateEmail = 'mail already taken'
