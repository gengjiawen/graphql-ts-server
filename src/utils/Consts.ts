export const JWT_SECRET = 'winds of winter'
