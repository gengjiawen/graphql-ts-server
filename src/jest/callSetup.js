require('ts-node/register')
module.exports = require('./setup').default
