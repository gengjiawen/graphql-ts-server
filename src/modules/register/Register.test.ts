import { User } from '../../entity/User'
import { duplicateEmail } from './ErrorMessages'
import { createTypeormConn } from '../../utils/CreateTyepeormConn'
import { graphQLRequest, mapErrorsMessage } from '../../jest/GraphQLTestUtil'
import { registerMutation } from '../../jest/shared.query'

beforeAll(async () => {
  await createTypeormConn()
})

const email = 'jack_register@test.com'
const password = '123456'

test('register user', async () => {
  const response: any = await graphQLRequest(registerMutation(email, password))
  expect(response.data.data.register.token).not.toBeNull()
  const users = await User.find({ where: { email } })
  expect(users).toHaveLength(1)
  const user = users[0]
  expect(user.email).toEqual(email)
  expect(user.password).not.toEqual(password)

  // test duplicate email
  const response2 = await graphQLRequest(registerMutation(email, password))
  expect(response2.data.data).toBeNull()
  expect(response2.data.errors[0].message).toEqual(duplicateEmail)

  // catch bad password
  const response3 = await graphQLRequest(registerMutation(email, 'ab'))
  expect(
    mapErrorsMessage(response3)[0].validationErrors.map((i: any) => i.constraints)
  ).toMatchSnapshot('catch bad password')

  // catch bad email and password
  const response4 = await graphQLRequest(registerMutation('ab', 'cd'))
  expect(
    mapErrorsMessage(response4)[0].validationErrors.map((i: any) => i.constraints)
  ).toMatchSnapshot('catch bad email and password')
})
