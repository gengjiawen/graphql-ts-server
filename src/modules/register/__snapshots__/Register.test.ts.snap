// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register user: catch bad email and password 1`] = `
Array [
  Object {
    "isEmail": "email must be an email",
  },
  Object {
    "length": "password must be longer than or equal to 3 characters",
  },
]
`;

exports[`register user: catch bad password 1`] = `
Array [
  Object {
    "length": "password must be longer than or equal to 3 characters",
  },
]
`;
